File: week1/0142. Linked List Cycle II/linked-list-cycleII.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None
# 快慢指针变形
class Solution:
    def detectCycle(self, head: ListNode) -> ListNode:
        slow, fast = head, head
        while slow and fast and fast.next:
            slow = slow.next
            fast = fast.next.next
            if slow == fast:
                # 查找交叉点
                dection = head
                while dection != slow:
                    slow = slow.next
                    dection = dection.next
                return dection
        return None

File: week1/0142. Linked List Cycle II/linked-list-cycleII.go
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
 func detectCycle(head *ListNode) *ListNode {
    slow, fast := head, head
    for slow != nil && fast != nil && fast.Next != nil{
        slow = slow.Next
        fast = fast.Next.Next
        if slow == fast{
            dection := head
            for dection != slow{
                dection = dection.Next
                slow = slow.Next
            }
            return dection
        }
    }
    return nil
    
}